iotc-azrtos-sdk/authentication/driver/to_auth_driver.c

```diff
@@ -0,0 +1,424 @@
+#define DC_MAGIC 0x3c
+
+struct to_driver_context {
+    struct to_driver_parameters driver_parameters;
+    IotcAzccCryptoConfig crypto_config;
+    TOSE_ctx_t* to_ctx;
+    uint8_t cert_buffer[TO_CERT_X509_MAXSIZE];
+    char cn_buffer[TO_CERT_SUBJECT_CN_MAXSIZE + 1];
+    char magic;
+};
+
+#define ASN1_TAG_SET 0x31
+#define ASN1_TAG_LEN_SIZE 2
+#define ASN1_TAG_SEQ 0x30
+#define ASN1_TAG_STR 0x0C
+static const uint8_t ASN1_OID_CN[] = {0x06, 0x03, 0x55, 0x04, 0x03};
+
+static TO_ret_t build_dn_from_cn(uint8_t *dn, const char* cn, uint16_t *size)
+{
+    if ((dn == NULL) || (cn == NULL) || (size == NULL))
+    {
+        return TO_INVALID_PARAM;
+    }
+    uint32_t cn_len = strlen(cn);
+    if (cn_len > TO_CERT_SUBJECT_CN_MAXSIZE)
+    {
+        return TO_INVALID_PARAM;
+    }
+    uint8_t total_len = (uint8_t)(ASN1_TAG_LEN_SIZE + sizeof(ASN1_OID_CN) + ASN1_TAG_LEN_SIZE + cn_len);
+    uint8_t i = 0;
+    dn[i++] = ASN1_TAG_SET;
+    dn[i++] = total_len;
+    dn[i++] = ASN1_TAG_SEQ;
+    dn[i] = total_len - i + 1;
+    i++;
+    memcpy(&dn[i], ASN1_OID_CN, sizeof(ASN1_OID_CN));
+    i += sizeof(ASN1_OID_CN);
+    dn[i++] = ASN1_TAG_STR;
+    dn[i] = total_len - i + 1;
+    i++;
+    snprintf((char*)&dn[i], cn_len + 1, "%s", cn);
+    *size = ASN1_TAG_LEN_SIZE + total_len;
+
+    return TO_OK;
+}
+
+static bool is_context_valid(IotcAuthInterfaceContext context) {
+	struct to_driver_context *to_context = (struct to_driver_context*) context;
+	bool ret = DC_MAGIC == to_context->magic;
+	if (!ret) {
+		printf("ADTO: Context is invalid\r\n");
+	}
+	return ret;
+}
+
+static NX_CRYPTO_KEEP UINT  to_ecdsa_operation(UINT op,      /* Encrypt, Decrypt, Authenticate */
+        VOID *handle, /* Crypto handler */
+        struct NX_CRYPTO_METHOD_STRUCT *method,
+        UCHAR *key, // our context will be stored in the key
+        NX_CRYPTO_KEY_SIZE key_size_in_bits,
+        UCHAR *input,
+        ULONG input_length_in_byte,
+        UCHAR *iv_ptr,
+        UCHAR *output,
+        ULONG output_length_in_byte,
+        VOID *crypto_metadata,
+        ULONG crypto_metadata_size,
+        VOID *packet_ptr,
+        VOID (*nx_crypto_hw_process_callback)(VOID *, UINT))
+{
+	if (op == NX_CRYPTO_EC_CURVE_SET) {
+		return _nx_crypto_method_ecdsa_operation(op, handle, method, key, key_size_in_bits, input, input_length_in_byte,
+					iv_ptr, output, output_length_in_byte, crypto_metadata, crypto_metadata_size, packet_ptr,
+					nx_crypto_hw_process_callback);
+
+	} else if (op == NX_CRYPTO_AUTHENTICATE) {
+		if (!is_context_valid(key)) return NX_CRYPTO_NOT_SUCCESSFUL;
+		struct to_driver_context* to_context = (struct to_driver_context*) key;
+		static uint8_t s_r_buffer[SR_BUFFER_SIZE];
+		TO_ret_t ret = TOSE_sign_hash(to_context->to_ctx, to_context->driver_parameters.operational_identity_slot, input, s_r_buffer);
+		if (TORSP_SUCCESS != ret) {
+	    	printf("ADTO: TOSE_sign error 0x%04X\r\n", ret);
```

Comment:
tabs/spaces inconsistency

Reply:
Those tab/space issues tend to pile up as we keep forgetting to update editor preferences and we edit the same files using different IDEs from Microchip, NXP, STM and whatnot. I tend to set tabs to 4 spaces in all my editors so I don't spot those. Fixed with a commit.